### PR TITLE
Added filter nfinset in product search

### PR DESCRIPTION
**Pull request by ronak2ram**
### Description (*)
I added `nfinset` filter for the product search
https://github.com/magento/graphql-ce/blob/2.3-develop/lib/internal/Magento/Framework/DB/Adapter/Pdo/Mysql.php#L2947

### Manual testing scenarios (*)
Request: 
```
{
  products(
    pageSize: 10
    filter: {
      name:{finset:["Sprite Yoga Strap 6 foot"]}
      }
  )
  {
    total_count
    items {
      id
      name
      sku
      price {
        regularPrice {
          amount {
            value
            currency
          }
        }
      }
    }
    page_info {
      page_size
      current_page
    }
  }
}
```
Response:
```
{
  "data": {
    "products": {
      "total_count": 1,
      "items": [
        {
          "id": 1,
          "name": "Sprite Yoga Strap 6 foot",
          "sku": "24-WG085",
          "price": null
        }
      ],
      "page_info": {
        "page_size": 10,
        "current_page": 1
      }
    }
  }
}
```
Request:
```
{
  products(
    pageSize: 10
    filter: {
      name:{nfinset:["Sprite Yoga Strap 6 foot"]}
      }
  )
  {
    total_count
    items {
      id
      name
      sku
      price {
        regularPrice {
          amount {
            value
            currency
          }
        }
      }
    }
    page_info {
      page_size
      current_page
    }
  }
}
```
Response:
```
{
  "data": {
    "products": {
      "total_count": 5,
      "items": [
        {
          "id": 2,
          "name": "Sprite Yoga Strap 8 foot",
          "sku": "24-WG086",
          "price": null
        },
        {
          "id": 3,
          "name": "Sprite Yoga Strap 10 foot",
          "sku": "24-WG087",
          "price": null
        },
        {
          "id": 4,
          "name": "Set of Sprite Yoga Straps",
          "sku": "24-WG085_Group",
          "price": null
        },
        {
          "id": 5,
          "name": "Sprite Yoga Strap Dynamic Bundle",
          "sku": "24-WG085-bundle-dynamic",
          "price": null
        },
        {
          "id": 6,
          "name": "Sprite Yoga Strap Fixed Bundle",
          "sku": "24-WG085-bundle-fixed",
          "price": null
        }
      ],
      "page_info": {
        "page_size": 10,
        "current_page": 1
      }
    }
  }
}
```

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

File: app/code/Magento/GraphQl/etc/schema.graphqls [hunk 1]
```diff
@@ -10,6 +10,7 @@ type Mutation {
 input FilterTypeInput @doc(description: "FilterTypeInput specifies which action will be performed in a query ") {
     eq: String @doc(description: "Equals")
     finset: [String] @doc(description: "Find in set. The value can contain a set of comma-separated values")
+    nfinset: [String] @doc(description: "Not find in set. The value can contain a set of comma-separated values")
```
Comment (naydav) [inline app/code/Magento/GraphQl/etc/schema.graphqls]: Please, cover this case with API-functional tests
We can apply this filter for product
Thanks

**Review by naydav — CHANGES_REQUESTED** [review]

File: app/code/Magento/GraphQl/etc/schema.graphqls (hunk 1 above)
Comment (ronak2ram) [reply app/code/Magento/GraphQl/etc/schema.graphqls]: @naydav I added API-functional tests

Comment (naydav) [reply app/code/Magento/GraphQl/etc/schema.graphqls]: @ronak2ram 
I mean create need to create one more testcase in ``\Magento\GraphQl\Catalog\ProductViewTest`
1) preconditions: some fixture with products (can used existed)
2) perform query with `nfinset` condition
3) check that result is proper

Thanks

File: dev/tests/api-functional/testsuite/Magento/GraphQl/Catalog/ProductViewTest.php [hunk 2]
```diff
@@ -670,6 +670,30 @@ public function testProductPrices()
...
+                   id
+                  name
+                  sku
+               }
+           }
+       }
+QUERY;
+
+        $response = $this->graphQlQuery($query);
+        $this->assertNotEquals($response['products']['items'][0]['sku'], $productSku);
```
Comment (naydav) [inline dev/tests/api-functional/testsuite/Magento/GraphQl/Catalog/ProductViewTest.php]: Need to make sure that product is missed in full result
`self::assertNotContains($productSku, array_column($items, 'sku'))`

**Review by naydav — CHANGES_REQUESTED** [review]

**Comment by naydav** [comment]
@ronak2ram 
`nfinset: [String] @doc(description: "Not find in set. The value can contain a set of comma-separated values")`

Doesn't work with multiple values.
If `value can contain a set of comma-separated values` why do we need parameter as `[String]`(list) ?
Looks like it's a mistake in the description. The parameter should be a list of values (for excluding).

But if you pass a few values
`products(filter: {sku: {nfinset: ["search_product_1", "search_product_2"]}})`
it leads to MySQL error on server

**Comment by naydav** [comment]
@ronak2ram 
Basic on our scheme value should be a list of strings
```
finset: [String]
nfinset: [String]
```
We could not change `finset: [String]` to `finset: String` because it will be backward incompatible changes in the scheme. Also, it's more convenient to have a possibility to filter by a few values.

So, we need to implement support of functionality if it is not working now.
We can take a value as a list and convert it to `string` and perform a search by this condition.
If this case doesn't work in Magento need to create a separate ticket and describe the problem


Now we have an issue with `finset` functionality, it leads to MySQL error, if we merge PR as is we will get one more MySQL error with
```
products(filter: {sku: {nfinset: ["search_product_1", "search_product_2"]}})
```
request.

**Comment by ronak2ram** [comment]
@naydav thanks for the update.
I did research on `finset`. we can use `FIND_IN_SET` only for a single value search same happen in magento.
We can't search two value in a single FIND_IN_SET MySQL query.
You can refer this link for more details.
http://www.mysqltutorial.org/mysql-find_in_set/
https://github.com/magento/graphql-ce/blob/2.3-develop/lib/internal/Magento/Framework/DB/Adapter/Pdo/Mysql.php#L2946

**Comment by naydav** [comment]
> @naydav thanks for the update.
> I did research on `finset`. we can use `FIND_IN_SET` only for a single value search same happen in magento.
> We can't search two value in a single FIND_IN_SET MySQL query.
> You can refer this link for more details.
> http://www.mysqltutorial.org/mysql-find_in_set/
> https://github.com/magento/graphql-ce/blob/2.3-develop/lib/internal/Magento/Framework/DB/Adapter/Pdo/Mysql.php#L2946

Sorry, correct me if I am wrong, you can work with two or more values:
`FIND_IN_SET(column, 'x,y,z')` is the same as `column IN (x, y, z)`

So, if in GraphQL we have a request like ['simple-1', 'simple-2'] we need to convert the value to 'simple-1, simple-2'
@paliarush what is your opinion? Do we want to do this? or we need just to fix scheme for working with single value?

**Comment by paliarush** [comment]
Agree with the last comment from @naydav

File: lib/internal/Magento/Framework/DB/Adapter/Pdo/Mysql.php [hunk 3]
```diff
@@ -2976,6 +2976,9 @@ public function prepareSqlCondition($fieldName, $condition)
                 if (($key == 'in' || $key == 'nin') && is_string($value)) {
                     $value = explode(',', $value);
                 }
+                if (($key == 'finset' || $key == 'nfinset') && is_array($value)) {
+                    $value = implode(',', $value);
+                }
```
Comment (ronak2ram) [inline lib/internal/Magento/Framework/DB/Adapter/Pdo/Mysql.php]: @naydav @paliarush 
I added this for resolve issue.
can you please check it.

**Comment by naydav** [comment]
@TomashKhamlai, please, check this issue
Check case with possible SQL-injection
Thanks

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/graphql-ce?pullRequest=528) <br/>All committers have signed the CLA.

**Comment by VladimirZaets** [comment]
Hi @ronak2ram. Thanks for collaboration. We had a discussion about `finset` and `nfinset` and made decision deprecate these fields. We don't recommend to store data as comma-separated values, so we don't need these fields.

**Review by VladimirZaets — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @VladimirZaets, thank you for the review.
[ENGCOM-5346](https://jira.corp.magento.com/browse/ENGCOM-5346) has been created to process this Pull Request
:eight_spoked_asterisk: @VladimirZaets, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by keharper** [comment]
@ronak2ram would you open a devdocs PR against https://devdocs.magento.com/guides/v2.3/graphql/queries.html to remove finset?

**Pull request merged**